src/main/ApplicationImpl.cpp

```diff
@@ -190,105 +190,112 @@ maybeRebuildLedger(Application& app, bool applyBuckets)
     std::set<LedgerEntryType> toRebuild;
     auto& ps = app.getPersistentState();
     auto bucketListDBEnabled = app.getConfig().isUsingBucketListDB();
-    for (auto let : xdr::xdr_traits<LedgerEntryType>::enum_values())
+
+#ifdef BUILD_TESTS
+    if (app.getConfig().MODE_USES_IN_MEMORY_LEDGER)
```

Comment:
I'm a bit confused by this, in-memory mode shouldn't have anything to do with the rebuild logic here. It was added just to deal with buggy schema upgrades: https://github.com/stellar/stellar-core/pull/3031, so this shouldn't impact testing?

Reply:
This is necessary, as the new version of "in memory mode" actually does persist and commit to a SQL DB for the offers table. Perhaps `MODE_USES_IN_MEMORY_LEDGER` is a poor named flag now, but with new in memory mode keeps all state in memory like the original prod version did in addition to writing to a SQL offers table. Reason being there was no reason we couldn't forward offer related writes and queries to the actual SQL table and it gives us better coverage.

Reply:
In particular there were some test cases failing without this. I don't remember exactly which ones, but I think it was in the upgrades path.

Reply:
Sorry, I still find it confusing - if you persist offers, why does not rebuilding them break the state? Rebuilding should end up in the same state that you started with (I guess I'm just surprised that this code, which was supposed to be functionally a no-op, breaks something)

Reply:
`maybeRebuildLedger`  doesn't actually "rebuild" state as in populating SQL tables. All it does is drop the tables, and if rebuild==true, recreates an empty version of the table. This is the only function used for creating ledger based SQL table, so if we don't support in-memory mode here, in-memory mode never actually creates the offers table.